File: go/cmd/examples/basic_usage.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/siemens/winccua-graphql-libs/go/pkg/winccunified"
)

func main() {
	baseURL := os.Getenv("WINCCUA_URL")
	username := os.Getenv("WINCCUA_USERNAME")
	password := os.Getenv("WINCCUA_PASSWORD")

	if baseURL == "" || username == "" || password == "" {
		log.Fatal("Please set WINCCUA_URL, WINCCUA_USERNAME, and WINCCUA_PASSWORD environment variables")
	}

	client := winccunified.NewClient(baseURL, username, password)

	ctx := context.Background()

	fmt.Println("Connecting to WinCC Unified server...")
	if err := client.Connect(ctx); err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer client.Disconnect(ctx)

	fmt.Println("Connected successfully!")

	// Read tags
	fmt.Println("\n=== Reading Tags ===")
	tagNames := []string{"Silo1_Temperature", "Silo1_Pressure", "InvalidTag"}
	tags, err := client.ReadTags(ctx, tagNames)
	if err != nil {
		log.Printf("Failed to read tags: %v", err)
	} else {
		for _, tag := range tags {
			if tag.Error != nil {
				fmt.Printf("Tag: %s - Error: %s\n", tag.Name, tag.Error.Description)
			} else {
				fmt.Printf("Tag: %s = %s (Quality: %s, Time: %v)\n", 
					tag.Name, tag.Value, tag.Quality, tag.Timestamp.Format(time.RFC3339))
			}
		}
	}

	// Write tags
	fmt.Println("\n=== Writing Tags ===")
	tagsToWrite := []winccunified.TagWrite{
		{Name: "Silo1_Temperature", Value: "25.5"},
		{Name: "Silo1_Pressure", Value: "1.2"},
	}

	writeResults, err := client.WriteTags(ctx, tagsToWrite)
	if err != nil {
		log.Printf("Failed to write tags: %v", err)
	} else {
		for _, result := range writeResults {
			if result.Error != nil {
				fmt.Printf("Write failed for %s: %s\n", result.Name, result.Error.Description)
			} else {
				fmt.Printf("Successfully wrote to %s\n", result.Name)
			}
		}
	}

	// Browse hierarchy
	fmt.Println("\n=== Browsing Tags ===")
	browseResult, err := client.Browse(ctx, "")
	if err != nil {
		log.Printf("Failed to browse: %v", err)
	} else {
		if browseResult.Error != nil {
			fmt.Printf("Browse error: %s\n", browseResult.Error.Description)
		} else {
			fmt.Printf("Found %d items:\n", len(browseResult.Items))
			for i, item := range browseResult.Items {
				if i >= 5 {
					fmt.Printf("... and %d more items\n", len(browseResult.Items)-5)
					break
				}
				fmt.Printf("  - %s (Type: %s, Children: %d)\n", 
					item.Name, item.Type, item.ChildrenCount)
			}
		}
	}

	// Get active alarms
	fmt.Println("\n=== Active Alarms ===")
	alarms, err := client.GetActiveAlarms(ctx)
	if err != nil {
		log.Printf("Failed to get active alarms: %v", err)
	} else {
		if len(alarms) == 0 {
			fmt.Println("No active alarms")
		} else {
			fmt.Printf("Found %d active alarms:\n", len(alarms))
			for i, alarm := range alarms {
				if i >= 3 {
					fmt.Printf("... and %d more alarms\n", len(alarms)-3)
					break
				}
				if alarm.Error != nil {
					fmt.Printf("  - Error getting alarm: %s\n", alarm.Error.Description)
				} else {
					fmt.Printf("  - %s: %s (State: %s)\n", 
						alarm.Name, alarm.Text, alarm.State)
				}
			}
		}
	}

	// Get redundancy state
	fmt.Println("\n=== Redundancy State ===")
	redundancyState, err := client.GetRedundancyState(ctx)
	if err != nil {
		log.Printf("Failed to get redundancy state: %v", err)
	} else {
		if redundancyState.Error != nil {
			fmt.Printf("Redundancy state error: %s\n", redundancyState.Error.Description)
		} else {
			fmt.Printf("Master: %t, State: %s\n", redundancyState.IsMaster, redundancyState.State)
		}
	}

	// Read historical values
	fmt.Println("\n=== Historical Values ===")
	endTime := time.Now()
	startTime := endTime.Add(-1 * time.Hour)
	
	historical, err := client.ReadHistoricalValues(ctx, "Silo1_Temperature", startTime, endTime, 10)
	if err != nil {
		log.Printf("Failed to read historical values: %v", err)
	} else {
		if historical.Error != nil {
			fmt.Printf("Historical values error: %s\n", historical.Error.Description)
		} else {
			fmt.Printf("Found %d historical values for %s:\n", len(historical.Values), historical.Name)
			for i, value := range historical.Values {
				if i >= 3 {
					fmt.Printf("... and %d more values\n", len(historical.Values)-3)
					break
				}
				fmt.Printf("  %s: %s (Quality: %s)\n", 
					value.Timestamp.Format(time.RFC3339), value.Value, value.Quality)
			}
		}
	}

	fmt.Println("\nExample completed successfully!")
}

File: c/examples/basic_usage.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include "../include/wincc_unified.h"

int main(int argc, char* argv[]) {
    const char* base_url = getenv("GRAPHQL_HTTP_URL");
    const char* username = getenv("GRAPHQL_USERNAME");
    const char* password = getenv("GRAPHQL_PASSWORD");
    
    if (!base_url || !username || !password) {
        fprintf(stderr, "Please set GRAPHQL_HTTP_URL, GRAPHQL_USERNAME, and GRAPHQL_PASSWORD environment variables\n");
        fprintf(stderr, "You can source the setenv.sh script to set these variables\n");
        return 1;
    }
    
    printf("Creating WinCC Unified client...\n");
    wincc_client_t* client = wincc_client_new(base_url, username, password);
    if (!client) {
        fprintf(stderr, "Failed to create client\n");
        return 1;
    }
    
    printf("Connecting to server...\n");
    wincc_error_t* error = wincc_connect(client);
    if (error && error->error_code && strcmp(error->error_code, "0") != 0) {
        fprintf(stderr, "Connection failed: %s - %s\n", error->error_code, error->description);
        wincc_error_free(error);
        wincc_client_free(client);
        return 1;
    }
    
    printf("Connected successfully!\n\n");
    
    printf("=== Reading Tags ===\n");
    const char* tag_names[] = {"Meter_Input_Value", "Meter_Output_Value", "HMI_Tag_1"};
    wincc_tag_results_t* tag_results = wincc_read_tags(client, tag_names, 3);
    
    if (tag_results) {
        for (size_t i = 0; i < tag_results->count; i++) {
            wincc_tag_result_t* tag = &tag_results->items[i];
            if (tag->error) {
                printf("Tag: %s - Error: %s\n", tag->name, tag->error->description);
            } else {
                printf("Tag: %s = %s (Quality: %s)\n", tag->name, tag->value, tag->quality);
            }
        }
        wincc_tag_results_free(tag_results);
    }
    
    printf("\n=== Writing Tags ===\n");
    wincc_tag_write_t tags_to_write[] = {
        {"HMI_Tag_1", "25.5"},
        {"HMI_Tag_2", "1.2"}
    };
    
    wincc_write_results_t* write_results = wincc_write_tags(client, tags_to_write, 2);
    if (write_results) {
        for (size_t i = 0; i < write_results->count; i++) {
            wincc_write_result_t* result = &write_results->items[i];
            if (result->error) {
                printf("Write failed for %s: %s\n", result->name, result->error->description);
            } else {
                printf("Successfully wrote to %s\n", result->name);
            }
        }
        wincc_write_results_free(write_results);
    }
    
    printf("\n=== Browsing Tags ===\n");
    wincc_browse_results_t* browse_results = wincc_browse(client, NULL);
    if (browse_results) {
        if (browse_results->error) {
            printf("Browse error: %s\n", browse_results->error->description);
        } else {
            printf("Found %zu items:\n", browse_results->count);
            for (size_t i = 0; i < browse_results->count && i < 5; i++) {
                wincc_browse_item_t* item = &browse_results->items[i];
                printf("  - %s (Type: %s, Children: %zu)\n", 
                       item->name, item->type, item->children_count);
            }
            if (browse_results->count > 5) {
                printf("  ... and %zu more items\n", browse_results->count - 5);
            }
        }
        wincc_browse_results_free(browse_results);
    }
    
    printf("\n=== Active Alarms ===\n");
    wincc_alarms_t* alarms = wincc_get_active_alarms(client);
    if (alarms) {
        if (alarms->count == 0) {
            printf("No active alarms\n");
        } else {
            printf("Found %zu active alarms:\n", alarms->count);
            for (size_t i = 0; i < alarms->count && i < 3; i++) {
                wincc_alarm_t* alarm = &alarms->items[i];
                if (alarm->error) {
                    printf("  - Error getting alarm: %s\n", alarm->error->description);
                } else {
                    printf("  - %s: %s (State: %s)\n", 
                           alarm->name, alarm->text, alarm->state);
                }
            }
            if (alarms->count > 3) {
                printf("  ... and %zu more alarms\n", alarms->count - 3);
            }
        }
        wincc_alarms_free(alarms);
    }
    
    printf("\nDisconnecting...\n");
    wincc_disconnect(client);
    wincc_client_free(client);
    
    printf("Done!\n");
    return 0;
}